src/libraries/System.Text.Json/src/System/Text/Json/Node/JsonObject.List.cs

```diff
@@ -0,0 +1,386 @@
+
+            JsonNode? existing = null;
+
+            if (_dictionary != null)
+            {
+                // Fast path if item doesn't exist in dictionary.
+                if (JsonHelpers.TryAdd(_dictionary, propertyName, node))
+                {
+                    node?.AssignParent(this);
+                    CreateDictionaryIfThreshold();
+                    _list.Add(new KeyValuePair<string, JsonNode?>(propertyName, node));
+                    return null;
+                }
+
+                existing = _dictionary[propertyName];
+                if (ReferenceEquals(existing, node))
+                {
+                    _lastKey = propertyName;
+                    _lastValue = node;
+
+                    // Ignore if the same value.
+                    return null;
+                }
+            }
+
+            int i = FindNodeIndex(propertyName);
+            if (i >= 0)
+            {
+                if (_dictionary != null)
+                {
+                    _dictionary[propertyName] = node;
+                }
+                else
+                {
+                    KeyValuePair<string, JsonNode?> current = _list[i];
+                    if (ReferenceEquals(current.Value, node))
+                    {
+                        // Ignore if the same value.
+                        return null;
+                    }
+
+                    existing = current.Value;
+                }
+
+                node?.AssignParent(this);
+                CreateDictionaryIfThreshold();
+                _list[i] = new KeyValuePair<string, JsonNode?>(propertyName, node);
+            }
+            else
+            {
+                node?.AssignParent(this);
+                CreateDictionaryIfThreshold();
+                _dictionary?.Add(propertyName, node);
+                _list.Add(new KeyValuePair<string, JsonNode?>(propertyName, node));
+                Debug.Assert(existing == null);
+            }
+
+            _lastKey = propertyName;
+            _lastValue = node;
+
+            return existing;
+        }
+
+        private int FindNodeIndex(string propertyName)
+        {
+            Debug.Assert(_list != null);
+
+            for (int i = 0; i < _list.Count; i++)
+            {
+                KeyValuePair<string, JsonNode?> current = _list[i];
+                if (GetStringComparer().Compare(propertyName, current.Key) == 0)
+                {
+                    return i;
+                }
+            }
+
+            return -1;
+        }
+
+        private void CreateDictionaryIfThreshold()
+        {
+            Debug.Assert(_list != null);
+
+            if (_dictionary == null && _list.Count > ListToDictionaryThreshold)
+            {
+                var dict = new Dictionary<string, JsonNode?>(GetStringComparer());
+
+                // 'foreach' has the advantage over a 'for(;;)' loop since it detects modifications.
+                foreach (KeyValuePair<string, JsonNode?> item in _list)
```

Comment:
Nit: couldn't `_list` be passed as an argument to the Dictionary constructor?

Reply:
For Core, yes, but that overload isn't present on NetFx and netstandard. I'll add a helper method with #ifdef to support that however.

Reply:
Good point. Perhaps it makes sense to keep as-is for the sake of readability.